### PR TITLE
chore(tailwind): make default foundation palette with css variables

### PR DESCRIPTION
**Pull request by mainframev**
[Slack message](https://skypicker.slack.com/archives/C05FDP4UM3N/p1697428801266619)
 Storybook: https://orbit-mainframev-chore-tailwind-adjustments-css-vars.surge.sh

### CODE REVIEW

File: packages/orbit-tailwind-preset/src/presets/foundation/getTailwindTheme.ts
Comment (DSil) [inline packages/orbit-tailwind-preset/src/presets/foundation/getTailwindTheme.ts:1]: This file is called `getTailwindTheme` but the function is called `getTailwindFoundation`. I'd rename the file to match the name of the function.

And are we supposed to maintain this file manually? Should it be documented?

File: packages/orbit-tailwind-preset/src/presets/components/index.ts [hunk 1]
```diff
@@ -66,191 +70,194 @@ const getBackgroundColors = () => {
...
   return {
     content: ["auto"],
-    presets: [orbitFoundationPreset],
+    presets: [orbitFoundationPreset(theme)],
     corePlugins: {
       preflight: disablePreflight ? false : undefined,
     },
     theme: {
+      variables: {
+        // paletteBlueNormal -> --palette-blue-normal
```
Comment (DSil) [inline packages/orbit-tailwind-preset/src/presets/components/index.ts:85]: This comment doesn't add much value to the code. It is, in fact, a comment of what the `kebabCase` function does, and not what this piece of code does. We either update it to the correct outputted format or we remove it from here.

File: packages/orbit-tailwind-preset/src/presets/components/index.ts [hunk 2]
```diff
@@ -66,191 +70,194 @@ const getBackgroundColors = () => {
...
+        DEFAULT: Object.entries(theme).reduce((acc, [key, value]) => {
```
Comment (DSil) [inline packages/orbit-tailwind-preset/src/presets/components/index.ts:86]: This feels like we're going round in circles. We have an object defined that matches keys to CSS vars. We create a new object with camelCase formatted tokens as keys and the same CSS vars as values. And then we iterate over the same object to convert the keys to kebab-case and keep the same values.

Can't a step be skipped here? 

I understand that maybe this is caused by the need of having some tailwind classes to use values directly and others to have CSS Vars, but maybe we can improve this. Colors to one side and everything else to the other might be a solution here. Since the only customizable thing in our themes are colors, we can use that in our favor

File: packages/orbit-tailwind-preset/src/presets/components/index.ts [hunk 3]
```diff
@@ -66,191 +70,194 @@ const getBackgroundColors = () => {
 
 const cfg = (options?: Options): Config => {
   const { disablePreflight = true } = options || {};
+  // make palette colors (foundation colors) defined as css vars and make components tokens inherit it
+  const theme = getTokens(cssVarsFoundation);
+  const componentTokens = getComponentLevelToken(theme);
```
Comment (DSil) [inline packages/orbit-tailwind-preset/src/presets/components/index.ts:75]: This is also hard to read. The function `getComponentLevelToken`, on its definition, names its parameter as `tokens`, but we're passing it a variable called `theme` (This was mentioned above). But then this function returns another function, it doesn't return a set of _componentTokens_ (which is what I infer when I read the variable name) that is then expected to be called, as we're doing. Please rename it. We can't call it again `getComponentLevelToken`, but then what do these two functions do differently?

File: packages/orbit-tailwind-preset/src/presets/foundation/cssVarsFoundation.ts
Comment (DSil) [inline packages/orbit-tailwind-preset/src/presets/foundation/cssVarsFoundation.ts:1]: How is this file generated? Do we need to maintain it manually?

**Review by DSil — CHANGES_REQUESTED** [review]
Overall good, some improvements needed

File: packages/orbit-tailwind-preset/src/index.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import orbitComponentsPreset from "./presets/components";
-import defaultFoundation from "./presets/foundation/theme/defaultFoundation";
+import defaultFoundation from "./presets/foundation/cssVarsFoundation";
```
[the remark this answers is not in the document]
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/index.ts:2]: that's why I kept that, exactly because of breaking change

File: packages/orbit-tailwind-preset/src/presets/components/index.ts [hunk 5]
```diff
@@ -66,191 +70,194 @@ const getBackgroundColors = () => {
 
 const cfg = (options?: Options): Config => {
   const { disablePreflight = true } = options || {};
+  // make palette colors (foundation colors) defined as css vars and make components tokens inherit it
+  const theme = getTokens(cssVarsFoundation);
```
[the remark this answers is not in the document]
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/presets/components/index.ts:74]: wdym by refactoring soon after this? there is no way how to resolve it without `getTokens` function that is documented in tokens package and basically the only way for theming available. By "refactor soon after it" I guess you mean the idea of having it generated from tokens package? The idea is good, but that won't be easy :)

File: packages/orbit-tailwind-preset/src/presets/foundation/cssVarsFoundation.ts
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/presets/foundation/cssVarsFoundation.ts:1]: It's not generated, yes, manually. Removed after suggestions from Boris of the need to have "weird mappers" and keep the thing simple.

File: packages/orbit-tailwind-preset/src/presets/foundation/getTailwindTheme.ts
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/presets/foundation/getTailwindTheme.ts:1]: nope, it should not. It's the same as before, I removed those folders for each object in the tailwind theme, like the folder for font, box-shadow, etc. And simplified that to a single object everything.

File: packages/orbit-tailwind-preset/src/presets/components/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (DSil) [reply packages/orbit-tailwind-preset/src/presets/components/index.ts:74]: This `getTokens` function is documented as a function to generate themes, which is ultimately what we're doing here. We're "creating a new theme" that is basically the same theme but with the tokens associated with CSS Vars instead of direct values. That seems very hacky to me. By "refactor soon after this" I mean multiple things:
- Don't have a function named `getTokens` that receives one schema and returns the merge of that schema with other default ones. That is simply not clear.
- Don't need to recreate the theme and iterate over the keys multiple times (this is related to what I commented [here](https://github.com/kiwicom/orbit/pull/4087#discussion_r1361821347) about going around in circles with object iterations)

Ofc I don't have a proposed solution for now, and I agree it's not straightforward, but I believe we can simplify or at least improve readability (and maintenance) here.

File: packages/orbit-tailwind-preset/src/presets/components/index.ts [hunk 6]
```diff
@@ -66,191 +70,194 @@ const getBackgroundColors = () => {
...
-          "heading-title2": defaultTokens.headingTitle2FontSize,
-          "heading-title3": defaultTokens.headingTitle3FontSize,
-          "heading-title4": defaultTokens.headingTitle4FontSize,
-          "heading-title5": defaultTokens.headingTitle5FontSize,
-          "heading-title6": defaultTokens.headingTitle6FontSize,
-          "button-large": defaultTokens.buttonLargeFontSize,
-          "button-normal": defaultTokens.buttonNormalFontSize,
-          "button-small": defaultTokens.buttonSmallFontSize,
-          "form-element-normal": defaultTokens.formElementNormalFontSize,
+          "heading-display": theme.headingDisplayFontSize,
```
[the remark this answers is not in the document]
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/presets/components/index.ts:112]: no. We making our theme preset themable, for now only inside (by calling getTokens), but later it's still has to have an option to do that froom the outside (the same way as with OrbitProvider now)
for Balkan and other projects. 

Plus, when now we have OrbitProvider and css variables changes, it might be needed to have custom foundation with custom typography

TLDR -> in components never should be used defaultTokens, only in foundation preset

File: packages/orbit-tailwind-preset/src/presets/components/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/presets/components/index.ts:74]: Yep, it can be solved that way only in the tokens package, here it's needed unfortunately to have it like that. I am trying to simplify this as much as possible for a minimal working solution.

[the remark this answers is not in the document]
Comment (DSil) [reply packages/orbit-tailwind-preset/src/presets/components/index.ts:74]: Yes, yes, I am not in favor of refactoring it now, let's not complicate it more. But please gather the ideas / feedback you have and the comments we have here so that we can consider a major refactor in the not-so-distant future

File: packages/orbit-tailwind-preset/src/presets/components/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (DSil) [reply packages/orbit-tailwind-preset/src/presets/components/index.ts:112]: Ok, I understand the concept of not calling defaultTokens on the components scope, but ultimately that's not necessary right now, or is it? I mean, would it break if we kept calling defaultTokens? Or would it just be wrong on the abstraction of the tokens?

File: packages/orbit-tailwind-preset/src/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/index.ts:2]: nope, not the same, but still should work for the users, as there is fallback to the default theme

[the remark this answers is not in the document]
Comment (DSil) [reply packages/orbit-tailwind-preset/src/index.ts:2]: Fine by me 😄 just document it as a breaking change and on the migration guide we mention that it should not impact anyone

File: packages/orbit-tailwind-preset/src/presets/components/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/presets/components/index.ts:112]: I think it will break it with *OrbitProvider*, when they change the theme object. Anyway, I feel like it's the right way and safer to have it like that, plus it's already preparing it for future adjustment 🤔

File: packages/orbit-tailwind-preset/src/presets/foundation/index.ts [hunk 7]
```diff
@@ -1,35 +1,46 @@
...
-      zIndex: {
-        default: String(defaultTokens.zIndexDefault),
-        sticky: String(defaultTokens.zIndexSticky),
-        "navigation-bar": String(defaultTokens.zIndexNavigationBar),
-        modal: String(defaultTokens.zIndexModal),
-        overlay: String(defaultTokens.zIndexModalOverlay),
-        drawer: String(defaultTokens.zIndexDrawer),
-        onTop: String(defaultTokens.zIndexOnTheTop),
+const config = (tokens: typeof defaultTokens): Config => {
+  const theme = getTailwindFoundation(tokens);
```
[the remark this answers is not in the document]
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/presets/foundation/index.ts:7]: I renamed that to `getTailwindTheme` to bring less confusion :)

File: packages/orbit-tailwind-preset/src/presets/components/index.ts (hunk 2 above)
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/presets/components/index.ts]: *Agreed in DMs, that it will be postponed on further larger refactoring*

File: packages/orbit-tailwind-preset/src/presets/foundation/getTailwindTheme.ts
Comment (DSil) [reply packages/orbit-tailwind-preset/src/presets/foundation/getTailwindTheme.ts:1]: Right. I meant internal documentation. But that's something we can do for the workshop

File: packages/orbit-tailwind-preset/src/presets/components/index.ts [hunk 8]
```diff
@@ -286,15 +279,29 @@ const cfg = (options?: Options): Config => {
...
-          const name = key.replace("text-", "");
-          return { ...acc, [name]: value };
-        }, {}),
-        backgroundColor: getBackgroundColors(tokens),
+        textColor: {
+          ...Object.entries(getForegroundColors(tokens)).reduce((acc, [key, value]) => {
+            const name = key.replace("text-", "");
+            return { ...acc, [name]: value };
+          }, {}),
+          // overrides for specific WL
```
Comment (DSil) [inline packages/orbit-tailwind-preset/src/presets/components/index.ts:287]: Are these overrides? Or are these just class definitions? What are we overriding?

"overrides for specific WL" is also a bit vague. What speficic?

**Review by DSil — COMMENTED** [review]
Just a comment about a comment 😄 We could also work on improving our internal documentation.

Please arrange commits before approval (there's a commit without scope and there are commits that introduce changes that are then removed later, so those can be squashed)

File: packages/orbit-tailwind-preset/src/presets/components/index.ts (hunk 8 above)
Comment (mainframev) [reply packages/orbit-tailwind-preset/src/presets/components/index.ts:287]: Idk what specific, as the document does not contain tokens per each WL, it just mentions which are used for overrides. Anyway, I made them at least available. Imo we shouldn't even allow usage of this tokens, as the WL need only the palette, but this hacks are used because they have wrongly defined their palettes or not constrast enough.

**Review by DSil — APPROVED** [review]
